rskj-core/src/main/java/co/rsk/pcc/bto/BTOUtils.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2019 RSK Labs Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.pcc.bto;
+
+import co.rsk.config.RskSystemProperties;
+import co.rsk.core.RskAddress;
+import co.rsk.pcc.NativeContract;
+import co.rsk.pcc.NativeMethod;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Precompiled contract that provides certain BTO (Bitcoin Token Offering)
+ * related utility functions.
+ *
+ * @author Ariel Mendelzon
+ */
+public class BTOUtils extends NativeContract {
+    private final BTOUtilsHelper helper;
+
+    public BTOUtils(RskSystemProperties config, RskAddress contractAddress) {
+        super(config, contractAddress);
+        this.helper = new BTOUtilsHelper();
+    }
+
+    @Override
+    public List<NativeMethod> getMethods() {
+        return Arrays.asList(
```

Comment:
This returns objects in an inconsistent state if the `init` method hasn't been called. I think this  refactor would fix this: pass the execution environment when calling `NativeMethod::execute`, instead of passing it at construction time.

Reply:
As is the case for the `Bridge` and `Remasc` contracts nowadays. At the moment a call to `init` before a call to `execute` is guaranteed throughout the possible usages.

To get a bigger understanding of certain design decisions, the idea here was to build a framework that would make it easy(er) to write precompiled contracts and not end up with massive cluttered files, while at the same time generating minimum friction with what currently exists. 

We could at some point in the future look at improving these, but that would require a more in-depth look.

Reply:
OK, I understand the limitations. Then how about checking this precondition and preventing an invalid state? I'm thinking something like:

```java
ExecutionEnvironment executionEnvironment = Objects.requireNotNull(getExecutionEnvironment());
return Arrays.asList(
    new ToBase58Check(executionEnvironment),
    ...
```

Or even better, it could be checked by:
* the `NativeMethod` constructor -> `InvalidArgumentException` and/or
* the `getExecutionEnvironment` method -> `InvalidStateException`

Checking it in the method would be the closest to the principle of failing fast, and it could also make sense to check the precondition in the constructor.

Reply:
We are checking the precondition in `NativeContract::execute`, I could add it to `NativeContract::getGasForData` as well.